src/container/cli.c

```diff
@@ -23,21 +23,35 @@
 #include "cli_internal.h"
 #include "rpc.h"
 
+int	dc_cont_proto_version;
+
 /**
  * Initialize container interface
  */
 int
 dc_cont_init(void)
 {
-	int rc;
+	int		rc;
+	uint32_t	ver_array[2] = {DAOS_CONT_VERSION - 1, DAOS_CONT_VERSION};
 
-	/* TODO: issue a cart protocol query to an engine, then register either the
-	 * latest, or latest-1 version of the container RPC protocol. See dc_pool_init().
-	 */
-	rc = daos_rpc_register(&cont_proto_fmt_v7, CONT_PROTO_CLI_COUNT,
-				NULL, DAOS_CONT_MODULE);
+	dc_cont_proto_version = 0;
+	rc = daos_rpc_proto_query(cont_proto_fmt_v6.cpf_base, ver_array,
+				  2, &dc_cont_proto_version);
+	if (rc)
+		return rc;
+
+	if (dc_cont_proto_version == DAOS_CONT_VERSION - 1) {
+		rc = daos_rpc_register(&cont_proto_fmt_v6, CONT_PROTO_CLI_COUNT,
+				       NULL, DAOS_CONT_MODULE);
+	} else if (dc_cont_proto_version == DAOS_CONT_VERSION) {
+		rc = daos_rpc_register(&cont_proto_fmt_v7, CONT_PROTO_CLI_COUNT,
+				       NULL, DAOS_CONT_MODULE);
+	} else {
+		D_ASSERT(0);
```

Comment:
Is `D_ASSERT()` the correct action here? What does this look like from the user perspective if they try to run with a version outside of these parameters? Seems like it would be better to log an error and return `-DER_PROTO` or perhaps a new `-DER_VER_INCOMPAT`, right?

Reply:
If run with a version outside of these parameters, daos_rpc_proto_query() will return error and won't come here. I could modify to return -DER_PROTO instead of assert as you wish.

Reply:
please see in daos_rpc_proto_query():
       for (i = 0; i < count; i++)
                if (ver_array[i] == rproto->version)
                        break;

        if (i == count) {
                D_ERROR("Invalid RPC protocol version %d\n", rproto->version);
                rc = -DER_PROTO;
        } else {
                *ret_ver = rproto->version;
        }

Reply:
> If run with a version outside of these parameters, daos_rpc_proto_query() will return error and won't come here. I could modify to return -DER_PROTO instead of assert as you wish.

I just noticed that @mchaarawi made the same point earlier. I think it's best to return an error instead of asserting, in case the usage of the code changes in the future. And I see that you pushed an update to do this, thanks.